### PR TITLE
Embind: Make subclasses inherit class functions

### LINKED ISSUE #18722: Embind: Class function not available on subclasses
Please include the following in your bug report:

**Version of emscripten/emsdk:**
```
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.30 (cfe2bdfe2692457cb5f5770672f6e5ccb3ffc2f2)
clang version 16.0.0 (https://github.com/llvm/llvm-project 800f0f1546b2352ba42a4777149afb13cb874fcd)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: /home/jcmorin/jcmenv/aswf/OpenTimelineIO-JS-Bindings/emsdk/upstream/bin
```

**Problem**:

When binding classes using Embind, functions declared as class functions are not available in subclasses. For example, I have this code:

```c++
#include <string>

#include <emscripten/bind.h>

namespace ems = emscripten;

class A
{
public:
    A(){};

    static std::string from_json_string(std::string const& input)
    {
        return input;
    };
};

class B : public A
{
public:
    B(){};
};

class C : public B
{
public:
    C(){};
};

EMSCRIPTEN_BINDINGS(inheritance)
{
    ems::class_<A>("A").constructor<>().class_function(
        "from_json_string",
        &A::from_json_string);

    ems::class_<B, ems::base<A>>("B").constructor<>();

    ems::class_<C, ems::base<B>>("C").constructor<>();
}
```
and I compile it with `emcc -lembind -o lib.js bindings.cpp -s MODULARIZE=1`, and then run it in JS:
```js
const factory = require('./lib');

factory().then((lib) => {
    console.log(lib.A.from_json_string("From A"))
    console.log(lib.B.from_json_string("From B"))
    console.log(lib.C.from_json_string("From C"))
})
```
It raises:
```
From A
/home/jcmorin/jcmenv/aswf/OpenTimelineIO-JS-Bindings/repros/inheritance/lib.js:165
      throw ex;
      ^

TypeError: lib.B.from_json_string is not a function
    at /home/jcmorin/jcmenv/aswf/OpenTimelineIO-JS-Bindings/repros/inheritance/test.js:5:23

Node.js v19.5.0
```

We can see that that `A.from_json_string()` works but `B.from_json_string()` and `C.from_json_string()` don't work.

I was kind of expecting class function to be available in sub-classes... Is this a bug or just how it should work?

Thanks a lot!

### PR DESCRIPTION
**Pull request by JeanChristopheMorinPerso**
Attempt to fix #18722.

ES6 style subclasses automatically inherit their parent's static functions. ES5 style classes don't. But I thought it would be nice if embind classes behaved more like ES6 classes.

The approach I took is to keep track of child classes in `<BaseClass>.__derivedClasses`, and then apply the class functions on child classes in `_embind_register_class_class_function`.

I initially tried to apply all class functions in `_embind_register_class`, but the problem with that is that the functions on the base class are not yet resolved (it's set to the `unboundTypesHandler` function). So the child class would end up with a class function that is never resolved (which would raise `Cannot call <class> due to unbound types ...`).

I don't know if there is a better way to achieve that. I'm open to suggestions!

### CODE REVIEW

File: src/embind/embind.js [hunk 1]
```diff
@@ -2371,6 +2380,15 @@ var LibraryEmbind = {
         } else {
           proto[methodName].overloadTable[argCount-1] = func;
         }
+
+        if (classType.registeredClass.__derivedClasses) {
+          for (const derivedClass of classType.registeredClass.__derivedClasses) {
+            if (!derivedClass.constructor.hasOwnProperty(methodName)) {
+              derivedClass.constructor[methodName] = func;
```
Comment (JeanChristopheMorinPerso) [inline src/embind/embind.js:2387]: This doesn't consider overloads... I'm guessing I would need to?

Comment (brendandahl) [reply src/embind/embind.js:2387]: To match how C++ would work we probably should, but I don't think it's needed right now. It also doesn't look like overloading completely works for regular class methods right now anyway.

Adding a todo here would be good though.

File: src/embind/embind.js [hunk 2]
```diff
@@ -2063,6 +2063,15 @@ var LibraryEmbind = {
                                                   upcast,
                                                   downcast);
 
+        if (registeredClass.baseClass) {
+          // Keep track of class hierarchy. USed to allow sub-classes to inherit class functions.
```
Comment (brendandahl) [inline src/embind/embind.js:2067]: nit: `Used`

**Review by brendandahl — CHANGES_REQUESTED** [review]

File: src/embind/embind.js (hunk 1 above)
Comment (JeanChristopheMorinPerso) [reply src/embind/embind.js:2388]: I added a todo.

**Comment by JeanChristopheMorinPerso** [comment]
Hello, I addressed the comments. I was wondering if someone could potentially review this PR please? I'd really love to see this land on the main branch :)

**Comment by kripken** [comment]
@brendandahl do you have time to look at this?

**Review by brendandahl — APPROVED** [review]

**Comment by JeanChristopheMorinPerso** [comment]
Done @brendandahl !

**Comment by JeanChristopheMorinPerso** [comment]
Some tests failed but it doesn't seem related to my changes...

**Comment by sbc100** [comment]
Can you put "embind" in the PR description somewhere so its clear what this change is about?

**Comment by JeanChristopheMorinPerso** [comment]
@sbc100 I added embind in the PR title. Does that work?

**Pull request merged**